xilem/src/property_tuple.rs

```diff
@@ -0,0 +1,183 @@
+    fn build_properties(&self) -> Properties {
+        let mut props = Properties::new();
+        if let Some(prop) = self.0.clone() {
+            props.insert(prop);
+        }
+        props
+    }
+
+    fn rebuild_properties(&self, prev: &Self, target: &mut WidgetMut<impl Widget>) {
+        if self.0 != prev.0 {
+            if let Some(prop) = self.0.clone() {
+                target.insert_prop(prop);
+            } else {
+                target.remove_prop::<P0>();
+            }
+        }
+    }
+}
+
+// We expect to use the ${index} metavariable here once it's stable
+// https://veykril.github.io/tlborm/decl-macros/minutiae/metavar-expr.html
+macro_rules! impl_property_tuple {
+    ($($Type: ident, $idx: tt);+) => {
+
+        impl<$($Type,)+> PropertyTuple for ($(Option<$Type>,)+)
+            where $($Type: Property + PartialEq + Clone,)+
+        {
+            fn build_properties(&self) -> Properties {
+                let mut props = Properties::new();
+                $(
+                    if let Some(prop) = self.$idx.clone() {
+                        props.insert(prop);
+                    }
+                )+
+                props
+            }
+
+            fn rebuild_properties(&self, prev: &Self, target: &mut WidgetMut<impl Widget>) {
+                $(
+                    if self.$idx != prev.$idx {
+                        if let Some(prop) = self.$idx.clone() {
+                            target.insert_prop(prop);
+                        } else {
+                            target.remove_prop::<$Type>();
+                        }
+                    }
+                )+
+            }
+
+        }
+
+    };
+}
+
+// The (P0,) one-element tuple case is covered outside the macro,
+// for easier code editing.
+
+impl_property_tuple!(P0, 0; P1, 1);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7; P8, 8);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7; P8, 8; P9, 9);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7; P8, 8; P9, 9; P10, 10);
+impl_property_tuple!(P0, 0; P1, 1; P2, 2; P3, 3; P4, 4; P5, 5; P6, 6; P7, 7; P8, 8; P9, 9; P10, 10; P11, 11);
+
+// ---
+
+// We expect to use the ${index} metavariable here once it's stable
+// https://veykril.github.io/tlborm/decl-macros/minutiae/metavar-expr.html
+
+/// Macro used to declare a list of properties for an element.
+///
+/// The expected way to invoke this macro is:
+///
+/// ```ignore
+/// declare_property_tuple!(
+///     MyWidgetProps;
+///     MyWidget;
```

Comment:
You're declaring this for a view, not a widget. The use of the name `MyWidget` is I think somewhat misleading.
It is technically correct that we tend to give the views the same names as the widgets, but I don't think that this kind of "cleverness" is helpful for learning.